File: katas/java/src/main/java/interactive_coding_challenges/arrays_and_strings/UniqueChars.java
package interactive_coding_challenges.arrays_and_strings;

import java.util.HashSet;
import java.util.Set;

/**
 * PROBLEM:
 * Implement an algorithm to determine if a string has all
 * unique characters.
 *
 * @author Mike Lowe
 */
public final class UniqueChars {

  // don't want instances
  private UniqueChars() {
  }

  public static boolean hasUniqueChars(final String input) {
    if (input == null) {
      return false;
    }
    final Set<Character> foundChars = new HashSet<>();
    for (int i = 0; i < input.length(); i++) {
      final char c = input.charAt(i);
      if (foundChars.contains(c)) {
        return false;
      } else {
        foundChars.add(c);
      }
    }
    return true;
  }

}

File: katas/javascript/interactive-coding-challenges/src/UniqueChars.js
/**
 * PROBLEM:
 * Implement an algorithm to determine if a string has all
 * unique characters.
 */

'use strict';

const UniqueChars = str => {
  const counts = {};
  
  for (let i = 0, length = str.length; i < length; i++) {
    const char = str[i];
    if (counts.hasOwnProperty(char)) {
      return false;
    } else {
      counts[char] = i;
    }
  }

  return true;
}

module.exports = UniqueChars;
